nbconvert/preprocessors/execute.py

```diff
@@ -409,76 +357,53 @@ def _update_display_id(self, display_id, msg):
                 outputs[output_idx]['data'] = out['data']
                 outputs[output_idx]['metadata'] = out['metadata']
 
-    def _wait_for_reply(self, msg_id, cell=None):
-        # wait for finish, with timeout
-        while True:
-            try:
-                if self.timeout_func is not None and cell is not None:
-                    timeout = self.timeout_func(cell)
-                else:
-                    timeout = self.timeout
-
-                if not timeout or timeout < 0:
-                    timeout = None
-                msg = self.kc.shell_channel.get_msg(timeout=timeout)
-            except Empty:
-                self.log.error(
-                    "Timeout waiting for execute reply (%is)." % self.timeout)
-                if self.interrupt_on_timeout:
-                    self.log.error("Interrupting kernel")
-                    self.km.interrupt_kernel()
-                    break
-                else:
-                    try:
-                        exception = TimeoutError
-                    except NameError:
-                        exception = RuntimeError
-                    raise exception("Cell execution timed out")
-
-            if msg['parent_header'].get('msg_id') == msg_id:
-                return msg
-            else:
-                # not our reply
-                continue
+    def _get_timeout(self, cell):
+        if self.timeout_func is not None:
+            timeout = self.timeout_func(cell)
+        else:
+            timeout = self.timeout
+
+        if not timeout or timeout < 0:
+            timeout = None
+        return timeout
 
     def run_cell(self, cell, cell_index=0):
-        msg_id = self.kc.execute(cell.source)
+        from tornado import ioloop
+
+        timeout = self._get_timeout(cell)
+        interrupt_timeout = None
+        if self.interrupt_on_timeout:
+            interrupt_timeout = timeout
+            timeout += 5
+
+        out_msgs = []
+
         self.log.debug("Executing cell:\n%s", cell.source)
-        exec_reply = self._wait_for_reply(msg_id, cell)
+        try:
+            reply = self.kc.execute_interactive(
+                cell.source,
+                output_hook=out_msgs.append,
+                timeout=timeout,
+                interrupt_timeout=interrupt_timeout,
+                idle_timeout=self.iopub_timeout,
+                raise_on_no_idle=self.raise_on_iopub_timeout,
+            )
+        except ioloop.TimeoutError:
```

Comment:
Not a problem for now, but could this ioloop.TimeoutError -> TimeoutError happen in jupyter_protocol? Does that seem appropriate to you?

Reply:
Yup, I think it would be appropriate for the BlockingKernelClient to do that translation.